### PR TITLE
Introduce a way to validate cluster state during regression tests

### PR DESCRIPTION
**Pull request by asimrp**
It is often observed in CI that a test that leaves the cluster in an inconsistent state (e.g. primary-mirror pair is not in sync, or primary has not finished crash recovery) cause several following tests in the schedule to fail.  What's worse, the culprit test itself may be reported as passed because its validation criteria did not include the state of the cluster.  This has found to mislead debugging efforts, ultimately waste of time.

To make debugging CI failures more efficient, this patch adds `--validate-system=TEST` option to `pg_regress`.  It specifies the name of the test that `pg_regress` would run to validate cluster state before running an actual test from a schedule.  If the validation test fails, further testing is aborted.  The validation is performed before running a group of tests specified on one line in the schedule file. Alternatively, validation could be performed before every single test, but let's not get there until absolutely necessary.

When the cluster validation is found to fail, the culprit is in the previously run test group.

The new option is turned off by default, leaving developer workflow unchanged.  It is enabled only in CI jobs.

This patch is built on the ground work and analysis laid out in PR #9865 by @gfphoenix78.  In particular, Wu Hao found that the validation overhead is small enough to make the proposed patch viable.

A sample `pg_regress` with this change output looks as follows.  The command was:
```
make PGPORT=7000 EXTRA_REGRESS_OPTS=--validate-system=check_cluster -C src/test/regress installcheck-good
```
It generated this output, note the `test check_cluster lines`:
```
test check_cluster            ... ok (0.03 sec)
test namespace                ... ok (0.17 sec)  (diff:0.08 sec)
test check_cluster            ... ok (0.03 sec)
parallel group (16 tests):  init_privs drop_operator lock security_label groupingsets replica_identity object_address join_hash spgist collate gist gin privileges tablesample rowsecurity brin
     brin                     ... ok (27.78 sec)  (diff:0.11 sec)
     gin                      ... ok (4.69 sec)  (diff:0.07 sec)
     gist                     ... ok (4.08 sec)  (diff:0.08 sec)
     spgist                   ... ok (2.64 sec)  (diff:0.08 sec)
     privileges               ... ok (6.51 sec)  (diff:0.15 sec)
     init_privs               ... ok (0.27 sec)  (diff:0.07 sec)
     security_label           ... ok (1.07 sec)  (diff:0.08 sec)
     collate                  ... ok (2.92 sec)  (diff:0.10 sec)
     lock                     ... ok (0.82 sec)  (diff:0.08 sec)
     replica_identity         ... ok (1.76 sec)  (diff:0.08 sec)
     rowsecurity              ... ok (8.62 sec)  (diff:0.23 sec)
     object_address           ... ok (2.00 sec)  (diff:0.22 sec)
     tablesample              ... ok (8.50 sec)  (diff:0.09 sec)
     groupingsets             ... ok (1.31 sec)  (diff:0.10 sec)
     drop_operator            ... ok (0.43 sec)  (diff:0.07 sec)
     join_hash                ... ok (2.43 sec)  (diff:0.08 sec)
test check_cluster            ... FAILED

========================
 1 of 138 tests failed.
========================
```

### CODE REVIEW

**Comment by asimrp** [comment]
I would like to pair with someone to validate the pipeline config change proposed in this patch.  Anyone interested?

File: src/test/regress/pg_regress.c [hunk 1]
```diff
@@ -2287,6 +2306,10 @@ run_single_test(const char *test, test_function tfunc)
 		differ |= newdiff;
 	}
 
+	double diff_secs = 0;
```
Comment (ashwinstar) [inline src/test/regress/pg_regress.c:2309]: Missing the context what is all this code trying to do? Seems doesn't belong to this commit or PR for sure.

File: concourse/scripts/ic_gpdb.bash [hunk 2]
```diff
@@ -47,6 +47,10 @@ function _main() {
         exit 1
     fi
 
+    if [ -n "${VALIDATE_SYSTEM_TESTNAME}" ]; then
+	MAKE_TEST_COMMAND="EXTRA_REGRESS_OPTS=--validate-system=${VALIDATE_SYSTEM_TESTNAME} ${MAKE_TEST_COMMAND}"
```
Comment (ashwinstar) [inline concourse/scripts/ic_gpdb.bash:51]: How do we control this only triggers for regress and isolation2 tests and not for tests like plpython, walrep, fsync etc.... Seems any module which uses schedule file for specifying tests will trigger this validation currently? Do we really need that.

File: concourse/scripts/ic_gpdb.bash [hunk 3]
```diff
@@ -47,6 +47,10 @@ function _main() {
         exit 1
     fi
 
+    if [ -n "${VALIDATE_SYSTEM_TESTNAME}" ]; then
```
[the remark this answers is not in the document]
Comment (gfphoenix78) [reply concourse/scripts/ic_gpdb.bash:50]: My previous way is to set an environment variable in the pipeline and check the variable in the Makefile under regress and isolation2. So the option will not be used by other test groups.

**Comment by gfphoenix78** [comment]
> I would like to pair with someone to validate the pipeline config change proposed in this patch. Anyone interested?

@asimrp I'd like to pair with you if you have time.

File: concourse/scripts/ic_gpdb.bash (hunk 2 above)
Comment (asimrp) [reply concourse/scripts/ic_gpdb.bash:51]: That is intentional.  It does feels like an overkill to do the check for schedules other than regress and isolation* but if we are aiming to be paranoid, let's be consistent with the paranoia.  One goal I have in mind is to keep the change as simple as possible.  Adding further ability to control the behavior might bring back the complexity from PR #9865.  Let me think if we can add ability to selectively enable this check without introducing additional complexity.

File: src/test/regress/pg_regress.c (hunk 1 above)
Comment (asimrp) [reply src/test/regress/pg_regress.c:2309]: I added it originally to measure time taken by the validation test (check_cluster).  It causes the seconds to be emitted after the "ok":
```
test check_cluster            ... ok (0.03 sec)
test namespace                ... ok (0.17 sec)  (diff:0.08 sec)
```
Let me remove it to minimize diff with respect to upstream.

**Comment by asimrp** [comment]
Thank you @gfphoenix78!

**Comment by ashwinstar** [comment]
Thinking more, what if we take this a step further to simplify and wrote the C function in regress itself to perform validation instead of invoking SQL test for it. Given currently we only wish to validate gp_segment_configuration. So, avoids cluttering of test output plus should be much more faster. Thoughts.

**Comment by asimrp** [comment]
That's an interesting idea.  One way is to write a UDF in C that returns true or false depending on whether gp_segment_configuration table contains down segments.  Invoking this C UDF from pg_regress does not seem simpler than what we already have in this PR.

A simpler alternative could be a C function in pg_regress to read the flat file that FTS writes a copy of cluster configuration to.  No database access is involved at all.  If a down segment is found, testing can be halted.  One downside of this approach is we will have to expose symbols that are currently private to backend code - segment status and the cluster configuration filename.

**Comment by ashwinstar** [comment]
This is what I am proposing.. seems does the job (will need finishing touches, roughly coded)

```C
diff --git a/src/test/regress/pg_regress.c b/src/test/regress/pg_regress.c
index a5d0443ed32..9a468f1d90c 100644
--- a/src/test/regress/pg_regress.c
+++ b/src/test/regress/pg_regress.c
@@ -128,6 +128,8 @@ static int  success_count = 0;
 static int     fail_count = 0;
 static int     fail_ignore_count = 0;
 
+static bool halt_work = false;
+
 static bool directory_exists(const char *dir);
 static void make_directory(const char *dir);
 
@@ -1470,6 +1472,76 @@ psql_command(const char *database, const char *query,...)
        }
 }
 
+/*
+ * Issue a command via psql, connecting to the specified database
+ *
+ */
+#define BUFSIZE 512
+static bool
+psql_command_output(const char *database, const char *query,...)
+{
+       char            query_formatted[1024];
+       char            query_escaped[2048];
+       char            psql_cmd[MAXPGPATH + 2048];
+       va_list         args;
+       char       *s;
+       char       *d;
+       char buf[BUFSIZE];
+       FILE *fp;
+
+       /* Generate the query with insertion of sprintf arguments */
+       va_start(args, query);
+       vsnprintf(query_formatted, sizeof(query_formatted), query, args);
+       va_end(args);
+
+       /* Now escape any shell double-quote metacharacters */
+       d = query_escaped;
+       for (s = query_formatted; *s; s++)
+       {
+               if (strchr("\\\"$`", *s))
+                       *d++ = '\\';
+               *d++ = *s;
+       }
+       *d = '\0';
+
+       /* And now we can build and execute the shell command */
+       snprintf(psql_cmd, sizeof(psql_cmd),
+                        "\"%s%spsql\" -X -t -c \"%s\" \"%s\"",
+                        bindir ? bindir : "",
+                        bindir ? "/" : "",
+                        query_escaped,
+                        database);
+
+       if ((fp = popen(psql_cmd, "r")) == NULL)
+       {
+               fprintf(stderr, _("error opening pipe!\n"), psql_cmd);
+               exit(2);
+       }
+
+       while(fgets(buf, BUFSIZE, fp) != NULL)
+       {
+               if (strcmp(buf, "\n") != 0)
+               {
+                       fprintf(stderr, "Cluster validation failed:\n%s\n", buf);
+                       return false;
+               }
+       }
+
+
+       if (pclose(fp))
+       {
+               fprintf(stderr, _("command exited with error status\n"), psql_cmd);
+               exit(2);
+       }
+
+       return true;
+}
+
+static bool
+validate_system()
+{
+       psql_command_output("postgres", "select dbid, content, status, role, preferred_role, mode from gp_segment_configuration where status = 'd' or preferred_role != role;");
+}
+
 /*
  * Spawn a process to execute the given shell command; don't wait for it
  *
@@ -2117,6 +2189,12 @@ run_schedule(const char *schedule, test_function tfunc)
                        exit(2);
                }
 
+               if (!validate_system())
+               {
+                       halt_work = true;
+                       break;
+               }
+
                gettimeofday(&start_time, NULL);
                if (num_tests == 1)
                {
@@ -3117,12 +3195,12 @@ regression_main(int argc, char *argv[], init_function ifunc, test_function tfunc
         */
        header(_("running regression test queries"));
 
-       for (sl = schedulelist; sl != NULL; sl = sl->next)
+       for (sl = schedulelist; sl != NULL && !halt_work; sl = sl->next)
        {
                run_schedule(sl->str, tfunc);
        }
 
-       for (sl = extra_tests; sl != NULL; sl = sl->next)
+       for (sl = extra_tests; sl != NULL && !halt_work; sl = sl->next)
        {
                run_single_test(sl->str, tfunc);
        }
```

**Comment by asimrp** [comment]
:+1: Thank you for the code snippet.  Indeed, this is even simpler, everything happens under the hood, no control to the user.  I like it.  @gfphoenix78 what do you think?  Should we replace this PR with Ashwin's version?

**Comment by gfphoenix78** [comment]
> 👍 Thank you for the code snippet. Indeed, this is even simpler, everything happens under the hood, no control to the user. I like it. @gfphoenix78 what do you think? Should we replace this PR with Ashwin's version?

It's much simpler, I agree.

**Comment by ashwinstar** [comment]
I pushed that rough patch here https://github.com/ashwinstar/gpdb/commit/3d85444ea2e9117f00a6039b1a2c033794221853 to help you easily iterate and take it to finish line whenever get to it.

**Comment by ashwinstar** [comment]
Closing this in favor of https://github.com/greenplum-db/gpdb/pull/11055
